internal/vm/task.go

```diff
@@ -0,0 +1,338 @@
+	"github.com/firecracker-microvm/firecracker-containerd/internal/bundle"
+	proto "github.com/firecracker-microvm/firecracker-containerd/proto/grpc"
+)
+
+// IODirection represents the direction in which container related I/O should
+// flow between a FIFO and a vsock connection (the possible directions being
+// from FIFO to vsock or from vsock to FIFO). This is needed to parameterize
+// functions that operate on either Host or Guest (in which case I/O directions
+// for the same stream are inverses) and functions that operate on stdin vs.
+// stdout/stderr (which again have inverse directions).
+type IODirection bool
+
+const (
+	// FIFOtoVSock represents I/O that should be copied from a FIFO to a VSock.
+	FIFOtoVSock IODirection = true
+
+	// VSockToFIFO represents I/O that should be copied from a VSock to a FIFO.
+	VSockToFIFO IODirection = false
+)
+
+func (d IODirection) opposite() IODirection {
+	return IODirection(!bool(d))
+}
+
+// VSockConnector is a function used as a callback for obtaining a vsock connection.
+// Implementations may include one that dials a remote vsock listener and one that
+// listens and accepts incoming connections.
+type VSockConnector func(ctx context.Context, port uint32) (net.Conn, error)
+
+// TaskManager manages a mapping of containerIDs to metadata for containerd tasks
+// being executed via a firecracker-containerd runtime. It's intended to be
+// abstracted over whether it's being executed on the Host or inside a VM Guest.
+type TaskManager interface {
+	AddTask(context.Context, string, taskAPI.TaskService, bundle.Dir, *proto.ExtraData, *cio.FIFOSet) (*Task, error)
+	Task(string) (*Task, error)
+	TaskCount() uint
+	Remove(context.Context, string)
+	RemoveAll(context.Context)
+}
+
+// NewTaskManager initializes a new TaskManager
+func NewTaskManager() TaskManager {
+	return &taskManager{
+		tasks: make(map[string]*Task),
+	}
+}
+
+type taskManager struct {
+	mu    sync.RWMutex
+	tasks map[string]*Task
+}
+
+// AddTask registers a task with the provided metadata with the taskManager.
+// taskService should implement the TaskService API for the task (i.e. Create, Kill, Exec, etc.).
+func (m *taskManager) AddTask(
+	ctx context.Context,
+	containerID string,
+	taskService taskAPI.TaskService,
+	bundleDir bundle.Dir,
+	extraData *proto.ExtraData,
+	fifoSet *cio.FIFOSet,
+) (*Task, error) {
+	err := bundleDir.Create() // no-op if it already exists
+	if err != nil {
+		return nil, err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.tasks[containerID]; ok {
+		return nil, errors.Errorf("cannot add duplicate task with container ID %q", containerID)
+	}
+
+	task := &Task{
+		TaskService: taskService,
+		ID:          containerID,
+
+		extraData: extraData,
+		bundleDir: bundleDir,
+		fifoSet:   fifoSet,
+
+		cancelCh: make(chan struct{}),
+	}
+
+	m.tasks[task.ID] = task
+	return task, nil
+}
+
+func (m *taskManager) get(containerID string) (*Task, error) {
+	m.mu.RLock()
```

Comment:
I think we should use a `sync.Map` or `RWLock` here, that way `get`s are not locking each other out.

Reply:
This is using an RWLock, it's calling `RLock()`

Reply:
Oh ffs. Totally missed the R lol.